cmd/libsnap-confine-private/snap-dir.c

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+
+#include "snap-dir.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "snap.h"  // For SC_SNAP_MOUNT_DIR_UNSUPPORTED
+
+static const char *_snap_mount_dir = NULL;
+
+const char *sc_snap_mount_dir(sc_error **errorp) {
+    sc_error *err = NULL;
+
+    if (_snap_mount_dir == NULL) {
+        err = sc_error_init_api_misuse("sc_probe_snap_mount_dir was not called yet");
+    }
+
+    sc_error_forward(errorp, err);
+    return _snap_mount_dir;
+}
+
+#define SC_PROC_1_ROOT "/proc/1/root"
+
+void sc_probe_snap_mount_dir(int root_fd, sc_error **errorp) {
+    char target[PATH_MAX];
+    struct stat sb;
+    ssize_t n;
+    sc_error *err = NULL;
+
+    const char *const probe_dir = SC_PROC_1_ROOT SC_PREFERRED_SNAP_MOUNT_DIR + (root_fd == AT_FDCWD ? 0 : 1);
+
+    // If /snap does not exist, then we assume the fallback directory is used.
+    if (fstatat(root_fd, probe_dir, &sb, AT_SYMLINK_NOFOLLOW) != 0) {
+        if (errno != ENOENT) {
+            err = sc_error_init_from_errno(errno, "cannot fstatat " SC_PROC_1_ROOT SC_PREFERRED_SNAP_MOUNT_DIR);
+            goto out;
+        }
+
+        _snap_mount_dir = SC_FALLBACK_SNAP_MOUNT_DIR;
+        return;
+    }
+
+    // If /snap exists it must be either a directory or a symbolic link.
+    switch (sb.st_mode & S_IFMT) {
+        case S_IFDIR:
+            _snap_mount_dir = SC_PREFERRED_SNAP_MOUNT_DIR;
+            break;
+
+        case S_IFLNK:
+            // If /snap exits and is a symbolic link it must point to the fallback directory.
+            n = readlinkat(root_fd, probe_dir, target, sizeof target);
+            if (n < 0 || n == sizeof(target)) {
+                err = sc_error_init_from_errno(errno, "cannot readlinkat " SC_PROC_1_ROOT SC_PREFERRED_SNAP_MOUNT_DIR);
+                goto out;
+            }
+
+            if (strncmp(target, SC_FALLBACK_SNAP_MOUNT_DIR, n) != 0) {
```

Comment:
you can use sc_streq

Reply:
I think we have a version with n but yeah.

Reply:
We don't have sc_streqn so I left this as-is. If you have strong feelings I can add one as well and use it.